File: frontend/src/types/export.ts
/**
 * Export Types
 *
 * Types for export job tracking and progress monitoring.
 * Supports background export jobs with real-time progress updates.
 *
 * @see backend/api/schemas/export.py - Backend Pydantic schemas
 * @see NEM-2385, NEM-2386
 */

// ============================================================================
// Enums
// ============================================================================

/**
 * Export job status values.
 *
 * Represents the lifecycle of an export job:
 * pending -> running -> completed | failed
 */
export type ExportJobStatus = 'pending' | 'running' | 'completed' | 'failed';

/**
 * Types of data exports available.
 */
export type ExportType = 'events' | 'alerts' | 'full_backup';

/**
 * Supported export file formats.
 */
export type ExportFormat = 'csv' | 'json' | 'zip' | 'excel';

// ============================================================================
// Request Types
// ============================================================================

/**
 * Available export column names.
 */
export type ExportColumnName =
  | 'event_id'
  | 'camera_name'
  | 'started_at'
  | 'ended_at'
  | 'risk_score'
  | 'risk_level'
  | 'summary'
  | 'detection_count'
  | 'reviewed'
  | 'object_types'
  | 'reasoning';

/**
 * Export column definition with field name and display label.
 */
export interface ExportColumnDefinition {
  /** Field name used in the export data */
  field: ExportColumnName;
  /** Human-readable display label */
  label: string;
  /** Description of the column */
  description: string;
}

/**
 * All available export columns with metadata.
 */
export const EXPORT_COLUMNS: ExportColumnDefinition[] = [
  { field: 'event_id', label: 'Event ID', description: 'Unique identifier for the event' },
  { field: 'camera_name', label: 'Camera', description: 'Name of the camera that captured the event' },
  { field: 'started_at', label: 'Started At', description: 'When the event started' },
  { field: 'ended_at', label: 'Ended At', description: 'When the event ended' },
  { field: 'risk_score', label: 'Risk Score', description: 'Numeric risk score (0-100)' },
  { field: 'risk_level', label: 'Risk Level', description: 'Risk level category (low/medium/high/critical)' },
  { field: 'summary', label: 'Summary', description: 'AI-generated summary of the event' },
  { field: 'detection_count', label: 'Detections', description: 'Number of detections in the event' },
  { field: 'reviewed', label: 'Reviewed', description: 'Whether the event has been reviewed' },
  { field: 'object_types', label: 'Object Types', description: 'Types of objects detected' },
  { field: 'reasoning', label: 'Reasoning', description: 'AI reasoning for the risk assessment' },
];

/**
 * Parameters for creating a new export job.
 */
export interface ExportJobCreateParams {
  /** Type of data to export */
  export_type?: ExportType;
  /** Output file format */
  export_format?: ExportFormat;
  /** Filter by camera ID */
  camera_id?: string | null;
  /** Filter by risk level */
  risk_level?: string | null;
  /** Filter events starting from this date (ISO format) */
  start_date?: string | null;
  /** Filter events ending before this date (ISO format) */
  end_date?: string | null;
  /** Filter by reviewed status */
  reviewed?: boolean | null;
  /** List of column field names to include (null for all) */
  columns?: ExportColumnName[] | null;
}

// ============================================================================
// Response Types
// ============================================================================

/**
 * Response when starting a new export job.
 */
export interface ExportJobStartResponse {
  /** Unique job identifier for tracking progress */
  job_id: string;
  /** Initial job status (always pending) */
  status: ExportJobStatus;
  /** Human-readable status message */
  message: string;
}

/**
 * Progress information for an export job.
 */
export interface ExportJobProgress {
  /** Total items to process (null if unknown) */
  total_items: number | null;
  /** Number of items processed so far */
  processed_items: number;
  /** Progress percentage (0-100) */
  progress_percent: number;
  /** Current processing step description */
  current_step: string | null;
  /** Estimated completion time (ISO format) */
  estimated_completion: string | null;
}

/**
 * Result information for a completed export.
 */
export interface ExportJobResult {
  /** Download path for the exported file */
  output_path: string | null;
  /** File size in bytes */
  output_size_bytes: number | null;
  /** Number of records exported */
  event_count: number;
  /** Export format used */
  format: string;
}

/**
 * Filter parameters used for an export job.
 */
export interface ExportFilterParams {
  /** Camera ID filter */
  camera_id: string | null;
  /** Risk level filter */
  risk_level: string | null;
  /** Start date filter (ISO format) */
  start_date: string | null;
  /** End date filter (ISO format) */
  end_date: string | null;
  /** Reviewed status filter */
  reviewed: boolean | null;
}

/**
 * Full export job response with status, progress, and result.
 */
export interface ExportJob {
  /** Unique export job identifier */
  id: string;
  /** Current job status */
  status: ExportJobStatus;
  /** Type of export */
  export_type: string;
  /** Export file format */
  export_format: string;
  /** Progress information */
  progress: ExportJobProgress;
  /** Job creation timestamp */
  created_at: string;
  /** Job start timestamp */
  started_at: string | null;
  /** Job completion timestamp */
  completed_at: string | null;
  /** Filter parameters used (JSON string, parse with parseFilterParams) */
  filter_params: string | null;
  /** Export result (populated when completed) */
  result: ExportJobResult | null;
  /** Error message (populated when failed) */
  error_message: string | null;
}

/**
 * Pagination metadata for list responses.
 */
export interface ExportPaginationMeta {
  /** Total number of items */
  total: number;
  /** Number of items per page */
  limit: number;
  /** Number of items skipped */
  offset: number;
  /** Cursor for next page (optional) */
  cursor: string | null;
  /** Cursor for next page (optional) */
  next_cursor: string | null;
  /** Whether there are more items */
  has_more: boolean;
}

/**
 * Paginated list of export jobs.
 */
export interface ExportJobListResponse {
  /** List of export jobs */
  items: ExportJob[];
  /** Pagination metadata */
  pagination: ExportPaginationMeta;
}

/**
 * Response when cancelling an export job.
 */
export interface ExportJobCancelResponse {
  /** Job ID that was cancelled */
  job_id: string;
  /** New job status after cancellation */
  status: ExportJobStatus;
  /** Cancellation status message */
  message: string;
  /** Whether cancellation was successful */
  cancelled: boolean;
}

/**
 * Export file download metadata.
 */
export interface ExportDownloadInfo {
  /** Whether the file is ready for download */
  ready: boolean;
  /** Exported filename */
  filename: string | null;
  /** MIME type of the file */
  content_type: string | null;
  /** File size in bytes */
  size_bytes: number | null;
  /** URL to download the file */
  download_url: string | null;
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Check if an export job is in a terminal state (completed or failed).
 */
export function isExportJobComplete(job: ExportJob): boolean {
  return job.status === 'completed' || job.status === 'failed';
}

/**
 * Check if an export job is currently running.
 */
export function isExportJobRunning(job: ExportJob): boolean {
  return job.status === 'running';
}

/**
 * Check if an export job is pending (waiting to start).
 */
export function isExportJobPending(job: ExportJob): boolean {
  return job.status === 'pending';
}

/**
 * Check if an export job has failed.
 */
export function isExportJobFailed(job: ExportJob): boolean {
  return job.status === 'failed';
}

/**
 * Format bytes to human-readable string.
 */
export function formatFileSize(bytes: number | null): string {
  if (bytes === null || bytes === 0) return '0 B';

  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  const base = 1024;
  const index = Math.floor(Math.log(bytes) / Math.log(base));
  const size = bytes / Math.pow(base, index);

  return `${size.toFixed(index > 0 ? 1 : 0)} ${units[index]}`;
}

/**
 * Calculate estimated time remaining based on progress.
 */
export function calculateTimeRemaining(
  startedAt: string | null,
  progressPercent: number
): string | null {
  if (!startedAt || progressPercent <= 0 || progressPercent >= 100) {
    return null;
  }

  const elapsed = Date.now() - new Date(startedAt).getTime();
  const estimatedTotal = (elapsed / progressPercent) * 100;
  const remaining = estimatedTotal - elapsed;

  if (remaining <= 0) return null;

  const minutes = Math.ceil(remaining / 60000);
  if (minutes === 1) return '~1 min remaining';
  if (minutes < 60) return `~${minutes} min remaining`;

  const hours = Math.ceil(minutes / 60);
  if (hours === 1) return '~1 hour remaining';
  return `~${hours} hours remaining`;
}

/**
 * Parse filter parameters from JSON string.
 */
export function parseFilterParams(filterParams: string | null): ExportFilterParams | null {
  if (!filterParams) return null;
  try {
    return JSON.parse(filterParams) as ExportFilterParams;
  } catch {
    return null;
  }
}

/**
 * Format filter parameters for display.
 */
export function formatFilterParams(filterParams: string | null): string[] {
  const params = parseFilterParams(filterParams);
  if (!params) return [];

  const filters: string[] = [];

  if (params.camera_id) {
    filters.push(`Camera: ${params.camera_id}`);
  }
  if (params.risk_level) {
    filters.push(`Risk: ${params.risk_level}`);
  }
  if (params.start_date) {
    filters.push(`From: ${new Date(params.start_date).toLocaleDateString()}`);
  }
  if (params.end_date) {
    filters.push(`To: ${new Date(params.end_date).toLocaleDateString()}`);
  }
  if (params.reviewed !== null) {
    filters.push(params.reviewed ? 'Reviewed only' : 'Unreviewed only');
  }

  return filters;
}

/**
 * Calculate export job duration in seconds.
 */
export function calculateDuration(startedAt: string | null, completedAt: string | null): number | null {
  if (!startedAt) return null;
  const endTime = completedAt ? new Date(completedAt).getTime() : Date.now();
  const startTime = new Date(startedAt).getTime();
  return Math.round((endTime - startTime) / 1000);
}

/**
 * Format duration for display.
 */
export function formatDuration(seconds: number | null): string {
  if (seconds === null) return '';
  if (seconds < 60) return `${seconds}s`;
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  if (minutes < 60) {
    return remainingSeconds > 0 ? `${minutes}m ${remainingSeconds}s` : `${minutes}m`;
  }
  const hours = Math.floor(minutes / 60);
  const remainingMinutes = minutes % 60;
  return `${hours}h ${remainingMinutes}m`;
}

File: backend/api/schemas/export.py
"""Pydantic schemas for export API endpoints.

This module defines request/response schemas for the export progress tracking
API. Supports background export jobs with real-time progress updates.

Export Types:
    - events: Security event data export
    - alerts: Alert history export
    - full_backup: Complete system backup

Export Formats:
    - csv: Comma-separated values
    - json: JSON format
    - zip: Compressed archive
    - excel: XLSX spreadsheet
"""

from datetime import datetime
from enum import StrEnum, auto

from pydantic import BaseModel, ConfigDict, Field

from backend.api.schemas.pagination import PaginationMeta


class ExportJobStatusEnum(StrEnum):
    """Export job status values."""

    PENDING = auto()
    RUNNING = auto()
    COMPLETED = auto()
    FAILED = auto()


class ExportTypeEnum(StrEnum):
    """Types of exports available."""

    EVENTS = "events"
    ALERTS = "alerts"
    FULL_BACKUP = "full_backup"


class ExportFormatEnum(StrEnum):
    """Export file formats."""

    CSV = "csv"
    JSON = "json"
    ZIP = "zip"
    EXCEL = "excel"


# =============================================================================
# Export Job Request Schemas
# =============================================================================


class ExportJobCreate(BaseModel):
    """Schema for creating an export job.

    Create a new background export job with optional filtering parameters.
    The job will be processed asynchronously and can be monitored via
    the job status endpoint.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "export_type": "events",
                "export_format": "csv",
                "camera_id": "front_door",
                "risk_level": "high",
                "start_date": "2025-01-01T00:00:00Z",
                "end_date": "2025-01-12T23:59:59Z",
                "reviewed": None,
            }
        }
    )

    export_type: ExportTypeEnum = Field(
        ExportTypeEnum.EVENTS,
        description="Type of data to export (events, alerts, full_backup)",
    )
    export_format: ExportFormatEnum = Field(
        ExportFormatEnum.CSV,
        description="Output file format (csv, json, zip, excel)",
    )

    # Filter parameters (for events export)
    camera_id: str | None = Field(None, description="Filter by camera ID")
    risk_level: str | None = Field(
        None,
        description="Filter by risk level (low, medium, high, critical)",
    )
    start_date: datetime | None = Field(
        None,
        description="Filter events starting from this date (ISO format)",
    )
    end_date: datetime | None = Field(
        None,
        description="Filter events ending before this date (ISO format)",
    )
    reviewed: bool | None = Field(
        None,
        description="Filter by reviewed status (true=reviewed, false=unreviewed, null=all)",
    )


class ExportJobStartResponse(BaseModel):
    """Response when creating an export job.

    Returns the job ID that can be used to track progress via
    GET /api/exports/{job_id}.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "job_id": "550e8400-e29b-41d4-a716-446655440000",
                "status": "pending",
                "message": "Export job created. Use GET /api/exports/{job_id} to track progress.",
            }
        }
    )

    job_id: str = Field(..., description="Unique job identifier for tracking progress")
    status: ExportJobStatusEnum = Field(
        ExportJobStatusEnum.PENDING,
        description="Initial job status (always pending)",
    )
    message: str = Field(..., description="Human-readable status message")


# =============================================================================
# Export Job Progress Schemas
# =============================================================================


class ExportJobProgress(BaseModel):
    """Schema for export job progress information.

    Detailed progress information for an export job, including
    timing, item counts, and current step.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "total_items": 1230,
                "processed_items": 245,
                "progress_percent": 20,
                "current_step": "Processing events...",
                "estimated_completion": "2025-01-12T14:35:00Z",
            }
        }
    )

    total_items: int | None = Field(None, description="Total items to process (null if unknown)")
    processed_items: int = Field(0, ge=0, description="Number of items processed so far")
    progress_percent: int = Field(0, ge=0, le=100, description="Progress percentage (0-100)")
    current_step: str | None = Field(None, description="Current processing step description")
    estimated_completion: datetime | None = Field(
        None,
        description="Estimated completion time (ISO format)",
    )


class ExportJobResult(BaseModel):
    """Schema for completed export job result.

    Information about the completed export, including download path
    and file statistics.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "output_path": "/api/exports/download/events_export_20250112_143000.csv",
                "output_size_bytes": 125432,
                "event_count": 1230,
                "format": "csv",
            }
        }
    )

    output_path: str | None = Field(None, description="Download path for the exported file")
    output_size_bytes: int | None = Field(None, ge=0, description="File size in bytes")
    event_count: int = Field(0, ge=0, description="Number of records exported")
    format: str = Field(..., description="Export format used")


# =============================================================================
# Export Job Response Schemas
# =============================================================================


class ExportJobResponse(BaseModel):
    """Schema for export job status response.

    Complete status information for an export job, including
    progress, timing, result, and any error information.
    """

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "status": "running",
                "export_type": "events",
                "export_format": "csv",
                "progress": {
                    "total_items": 1230,
                    "processed_items": 245,
                    "progress_percent": 20,
                    "current_step": "Processing events...",
                    "estimated_completion": "2025-01-12T14:35:00Z",
                },
                "created_at": "2025-01-12T14:30:00Z",
                "started_at": "2025-01-12T14:30:01Z",
                "completed_at": None,
                "result": None,
                "error_message": None,
            }
        },
    )

    id: str = Field(..., description="Unique export job identifier")
    status: ExportJobStatusEnum = Field(..., description="Current job status")
    export_type: str = Field(..., description="Type of export")
    export_format: str = Field(..., description="Export file format")

    # Progress information
    progress: ExportJobProgress = Field(
        default_factory=ExportJobProgress,
        description="Progress information",
    )

    # Timing information
    created_at: datetime = Field(..., description="Job creation timestamp")
    started_at: datetime | None = Field(None, description="Job start timestamp")
    completed_at: datetime | None = Field(None, description="Job completion timestamp")

    # Result information (populated on completion)
    result: ExportJobResult | None = Field(
        None,
        description="Export result (populated when completed)",
    )

    # Error information (populated on failure)
    error_message: str | None = Field(
        None,
        description="Error message (populated when failed)",
    )


class ExportJobListResponse(BaseModel):
    """Schema for export job list response with pagination."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "items": [
                    {
                        "id": "550e8400-e29b-41d4-a716-446655440000",
                        "status": "completed",
                        "export_type": "events",
                        "export_format": "csv",
                        "progress": {
                            "total_items": 1230,
                            "processed_items": 1230,
                            "progress_percent": 100,
                            "current_step": "Complete",
                        },
                        "created_at": "2025-01-12T14:30:00Z",
                        "started_at": "2025-01-12T14:30:01Z",
                        "completed_at": "2025-01-12T14:35:00Z",
                        "result": {
                            "output_path": "/api/exports/download/events_export_20250112_143000.csv",
                            "output_size_bytes": 125432,
                            "event_count": 1230,
                            "format": "csv",
                        },
                        "error_message": None,
                    }
                ],
                "pagination": {
                    "total": 1,
                    "limit": 50,
                    "offset": 0,
                    "cursor": None,
                    "next_cursor": None,
                    "has_more": False,
                },
            }
        }
    )

    items: list[ExportJobResponse] = Field(..., description="List of export jobs")
    pagination: PaginationMeta = Field(..., description="Pagination metadata")


# =============================================================================
# Export Job Update Schemas
# =============================================================================


class ExportJobUpdate(BaseModel):
    """Schema for updating an export job.

    Internal schema used by the export service to update job progress.
    Not exposed via API (jobs are managed internally).
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "status": "running",
                "processed_items": 500,
                "progress_percent": 40,
                "current_step": "Processing events...",
            }
        }
    )

    status: ExportJobStatusEnum | None = Field(None, description="New job status")
    total_items: int | None = Field(None, ge=0, description="Total items to process")
    processed_items: int | None = Field(None, ge=0, description="Items processed")
    progress_percent: int | None = Field(None, ge=0, le=100, description="Progress percentage")
    current_step: str | None = Field(None, description="Current step description")
    started_at: datetime | None = Field(None, description="Job start time")
    completed_at: datetime | None = Field(None, description="Job completion time")
    estimated_completion: datetime | None = Field(None, description="Estimated completion time")
    output_path: str | None = Field(None, description="Output file path")
    output_size_bytes: int | None = Field(None, ge=0, description="Output file size")
    error_message: str | None = Field(None, description="Error message if failed")


# =============================================================================
# Export Download Schemas
# =============================================================================


class ExportDownloadResponse(BaseModel):
    """Schema for export file download metadata.

    Returned when checking if a file is ready for download.
    The actual file is served via a separate streaming endpoint.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "ready": True,
                "filename": "events_export_20250112_143000.csv",
                "content_type": "text/csv",
                "size_bytes": 125432,
                "download_url": "/api/exports/550e8400-e29b-41d4-a716-446655440000/download",
            }
        }
    )

    ready: bool = Field(..., description="Whether the file is ready for download")
    filename: str | None = Field(None, description="Exported filename")
    content_type: str | None = Field(None, description="MIME type of the file")
    size_bytes: int | None = Field(None, ge=0, description="File size in bytes")
    download_url: str | None = Field(None, description="URL to download the file")


# =============================================================================
# Export Cancel Schema
# =============================================================================


class ExportJobCancelResponse(BaseModel):
    """Schema for export job cancellation response."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "job_id": "550e8400-e29b-41d4-a716-446655440000",
                "status": "failed",
                "message": "Export job cancelled by user",
                "cancelled": True,
            }
        }
    )

    job_id: str = Field(..., description="Job ID that was cancelled")
    status: ExportJobStatusEnum = Field(..., description="New job status after cancellation")
    message: str = Field(..., description="Cancellation status message")
    cancelled: bool = Field(..., description="Whether cancellation was successful")
